File: src/data_manipulation/remove_noise.jl
using Distributed

# Add the number of cores to use
addprocs(8)

@everywhere begin
    using BM3DDenoise
    using ImageIO
    using FileIO
    using Images
    using Glob
    using Printf

    # Function to denoise the image
    function denoise_image(img::AbstractArray, noise_variance::Float64)::AbstractArray
        img_array = float(channelview(img))  # Convert image to 2D array
        img_denoised = bm3d(img_array, noise_variance)  # Apply BM3D algorithm
        img_denoised = (img_denoised .- minimum(img_denoised)) ./ (maximum(img_denoised) - minimum(img_denoised)) # Normalize
        img_denoised = colorview(Gray, img_denoised)  # Convert back to grayscale image
        return img_denoised
    end

    # Function to process each file
    function process_file(file::String, noise_variance::Float64, base_path::String, base_path_modified::String)
        # Maintain the subfolder structure in the new path
        relative_path = relpath(file, base_path)
        save_path = joinpath(base_path_modified, relative_path)

        # Check if the file already exists in the output directory. If it does, then continue with the next iteration.
        if isfile(save_path)
            println("File already exists -> Skipping.")
            return
        end

        img = load(file)
        img_denoised = denoise_image(img, noise_variance)

        mkpath(dirname(save_path))
        save(save_path, img_denoised)
    end
end

function main()
    base_path = "data/CellData/OCT_white_to_black"
    base_path_modified = "data/CellData/OCT_mb3d"
    noise_variance = 0.15
    subfolders = [
        # "test/NORMAL",
        # "test/DRUSEN",
        # "test/DME",
        # "test/CNV",
        "train/DRUSEN",
        "train/NORMAL",
        "train/CNV",
        # "train/DME",
    ]

    println("Starting the main task")
    # Flatten the file structure
    all_files = [glob("*.jpeg", joinpath(base_path, subfolder)) for subfolder in subfolders]
    all_files = vcat(all_files...)  # Concatenate all the files into a single array

    # @sync @distributed for file in all_files
    #     process_file(file, noise_variance, base_path, base_path_modified)
    #     println("Finished processing $file")
    # end

    start_time = time()
    for (i, file) in enumerate(all_files)
        elapsed_time = time() - start_time
        hours, rem = divrem(elapsed_time, 3600)
        minutes, seconds = divrem(rem, 60)
        println("Time elapsed: $(floor(Int, hours))h $(floor(Int, minutes))m $(floor(Int, seconds))s")

        process_file(file, noise_variance, base_path, base_path_modified)

        println("Finished processing $file")
    end
end

main()

File: src/data_manipulation/remove_noise.py
import os
import time
import imageio
import numpy as np
from bm3d import bm3d, BM3DStages
from glob import glob

# Function to denoise the image
def denoise_image(img, noise_variance):
    img = img.astype(np.float64)/255  # normalize the data to 0 - 1
    # img_denoised = bm3d(img, noise_variance)
    img_denoised = bm3d(img, noise_variance, stage_arg=BM3DStages.ALL_STAGES)
    img_denoised = (img_denoised - np.min(img_denoised)) / (np.max(img_denoised) - np.min(img_denoised))  # Normalize
    img_denoised *= 255  # scale back to 0 - 255
    return img_denoised.astype(np.uint8)

# Function to process each file
def process_file(file, noise_variance, base_path, base_path_modified, counter):
    # Maintain the subfolder structure in the new path
    relative_path = os.path.relpath(file, base_path)
    save_path = os.path.join(base_path_modified, relative_path)

    # Check if the file already exists in the output directory. If it does, then continue with the next iteration.
    if os.path.isfile(save_path):
        print(f"File already exists -> Skipping: {file}")
        return counter

    img = imageio.imread(file)
    img_denoised = denoise_image(img, noise_variance)
    os.makedirs(os.path.dirname(save_path), exist_ok=True)
    imageio.imsave(save_path, img_denoised)

    return counter + 1


def main():
    base_path = "data/CellData/OCT_white_to_black"
    base_path_modified = "data/CellData/OCT_mb3d"
    noise_variance = 0.15
    subfolders = [
        "train/DRUSEN",
        "train/NORMAL",
        "train/CNV",
    ]

    print("Starting the main task")
    # Flatten the file structure
    all_files = [glob(os.path.join(base_path, subfolder, "*.jpeg")) for subfolder in subfolders]
    all_files = [file for sublist in all_files for file in sublist]  # flatten the list

    counter = 0
    start_time = time.time()
    for file in all_files:
        elapsed_time = time.time() - start_time
        hours, rem = divmod(elapsed_time, 3600)
        minutes, seconds = divmod(rem, 60)
        print(f"Counter {counter} Time: {int(hours)}h {int(minutes)}m {int(seconds)}s")

        counter = process_file(file, noise_variance, base_path, base_path_modified, counter)

        print(f"Finished processing {file}")

if __name__ == "__main__":
    main()
